### PR TITLE
fix(queue): prevent duplicate messages getting pushed to the queue

### PR DESCRIPTION
**Pull request by robfletcher**

### CODE REVIEW

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt [hunk 1]
```diff
@@ -58,17 +56,19 @@ class RedisQueue(
...
-            readMessage(id) { payload ->
-              callback.invoke(payload) {
-                ack(id)
-                fire<MessageAcknowledged>()
+        zrangeByScore(queueKey, 0.0, score(), 0, 1)
+          .firstOrNull()
+          ?.takeIf { id -> acquireLock(id) }
+          ?.also { id ->
+            readMessage(id) { message ->
+              callback.invoke(message) {
```
Comment (cfieber) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: this looks to me like the redis connection is held open while the callback handles the message, is that expected?

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt [hunk 2]
```diff
@@ -79,13 +79,15 @@ class RedisQueue(
 
   override fun push(message: Message, delay: TemporalAmount) {
     pool.resource.use { redis ->
-      val id = randomUUID().toString()
-      redis.multi {
-        hset(messagesKey, id, mapper.writeValueAsString(message))
-        zadd(queueKey, score(delay), id)
+      val messageHash = message.hashCode().toString()
```
Comment (cfieber) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: I think you need something other than hashCode here or you will get collisions

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt (hunk 1 above)
Comment (robzienert) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: Good catch. Looks that way to me as well.

Comment (robfletcher) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: That entire lambda ``` callback.invoke(message) {
  ackMessage(id)
}``` is called here https://github.com/spinnaker/orca/blob/master/orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/QueueProcessor.kt#L51-L65 so whatever the handler does is off in another thread.

I need to give this some thought.

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt (hunk 2 above)
Comment (robfletcher) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: Looking into using murmur

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt [hunk 3]
```diff
@@ -232,4 +250,13 @@ class RedisQueue(
...
     }
+
+  private fun JedisCommands.hgetInt(key: String, field: String, default: Int = 0) =
+    hget(key, field)?.toInt() ?: default
+
+  private fun Message.hash() =
+    Hashing
+      .murmur3_32()
+      .hashString(toString(), Charset.defaultCharset())
+      .toString()
```
Comment (robfletcher) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: Added a simple hash method. Could, of course, just use the `toString` value itself but this is simple enough.

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt [hunk 4]
```diff
@@ -58,66 +58,74 @@ class RedisQueue(
...
-              }
-            }
+      redis.zrangeByScore(queueKey, 0.0, score(), 0, 1)
+        .firstOrNull()
+        ?.takeIf { id -> redis.acquireLock(id) }
+        ?.also { id ->
+          val ack = this::ackMessage.curry(id)
+          redis.readMessage(id) { message ->
+            callback(message, ack)
           }
```
Comment (robfletcher) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: I think this clarifies how the callbacks work a little, at least I hope so.

`callback` is the lambda `QueueProcessor` uses to find a handler, dispatch the message (on another thread) and then call the `ack` function.

File: orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt (hunk 3 above)
Comment (robfletcher) [inline orca-queue-redis/src/main/kotlin/com/netflix/spinnaker/orca/q/redis/RedisQueue.kt]: For context the Kotlin data class `toString` contains all property values so it's unique.

**Pull request merged**